tsfc/spectral.py

```diff
@@ -68,28 +128,58 @@ def classify(argument_indices, expression):
         return COMPOUND
 
 
-def flatten(var_reps, index_cache):
-    assignments = unconcatenate([(variable, reduce(Sum, reps))
-                                 for variable, reps in var_reps],
-                                cache=index_cache)
+def delta_elimination(variable, sum_indices, args, rest):
+    """IndexSum-Delta cancellation for monomials."""
+    factors = list(args) + [variable, rest]  # construct factors
 
-    def group_key(assignment):
-        variable, expression = assignment
-        return variable.free_indices
+    def prune(factors):
+        result = remove_componenttensors(factors[:-1])
+        result.append(factors[-1])
+        return result
```

Comment:
During argument factorisation, the "atomics" are subexpressions of FInAT-provided tabulations, which means they are small so this index substitution is cheap. Moreover, `MonomialSum` collapses monomials with the same set of sum indices and atomics. Layers of `Indexed(ComponentTensor(..., ...), ...)` would atomics that are actually the same to look different, and thus would hinder the shortening of the `MonomialSum`.

The `rest` part, which is the last factor if you look at the ordering above, is the opposite of what is described above. It can be arbitrarily big, making `remove_componenttensors` costly, and there are not benefits to early removal of those `ComponentTensor`s.

Perhaps I should write some if this into a comment?